cpp/src/interpolate/cubic_spline.cu

```diff
@@ -63,11 +64,23 @@ struct parallel_search {
           RESULT_[index] = h+len - 2;
       });
   };
+  template <typename T>
+  std::enable_if_t<not std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
+                  cudf::column_view const& curve_ids,
+                  cudf::column_view const& prefixes,
+                  cudf::column_view const& query_coords,
+                  cudf::mutable_column_view const& result,
+                  rmm::mr::device_memory_resource *mr,
+                  cudaStream_t stream) {
+      CUDF_FAIL("Non-floating point operation is not supported.");
+  }
 };
 
 struct interpolate {
   template <typename T>
-  void operator()(cudf::column_view const& t,
+  std::enable_if_t<std::is_floating_point<T>::value, void>
+  operator()(cudf::column_view const& t,
```

Comment:
Not sure, but should these be `column_device_view and `table_device_view` ?

Reply:
I don't know. It didn't seem necessary or appropriate - I was able to verify that any data in a `column_view` I created was always in device memory, and ran into plenty of API issues whenever I used an explicit `device_view`

Reply:
I think this should just be a functor that gets called by thrust::transform, rather than a wrapper around thrust::transform...

Reply:
I tried to implement this with
```
template <typename T, std::enable_if_t<std::is_floating_point<T>::value>* = nullptr>
void operator()(...
```
if that's what you mean, but could only get it to compile this way.

Reply:
No that's not what I mean.

Reply:
What I mean is move the thrust call out of this functor. Make this functor just the body of the lambda that your thrust::for_Each calls. Then use this functor as the argument to the thrust::for_each...

High-level example.
Instead of this:
```class functor_old {
  operator()(...) {
    column out;
    thrust::for_each(input.begin(), input.end(), [out](auto in) { out[i] = in*2; });
}```

Do this

```class functor_new {
  operator()(T in) {
    return in*2; 
  };
};

// in the calling function that normally invokes functor(), do the thrust algorithm:
thrust::transform(input.begin(), input.end(), out.begin(), functor_new());
```

Reply:
Oh, sure, I can do that. But Lambda functions in C++ are so cool! So dreeaammy.

Reply:
Actually I don't know how to do this - I'm depending on the Lambda and the `thrust::for_each` to inject the various pointers used in the calculation, and the `index` of the current calculation, which identifies which spline among `offsets` I'm working on.

Reply:
So I tried my hand at simplifying this, and while I think I was able to convert it to c++ isms (haven't compiled / run), I'm not sure how much we gain in terms of clarity.
```
auto input_iter = thrust::make_zip_iterator(
    thrust::make_tuple(
        thrust::make_permutation_iterator(coefficients.column(0).data<T>(), result.data<T>()),
        thrust::make_permutation_iterator(coefficients.column(1).data<T>(), result.data<T>()),
        thrust::make_permutation_iterator(coefficients.column(2).data<T>(), result.data<T>()),
        thrust::make_permutation_iterator(coefficients.column(3).data<T>(), result.data<T>())
        t.data<T>())

thrust::transform(input_iter
                  input_iter + t.size(),
                  result.data<T>().get(),
                  __device__ (decltype(input_iter)::value_type input) {
                      T d0, d1, d2, d3, x;
                      std::tie(d0, d1, d2, d3, x) = input;
                      return d3 + x * (d2 + x * (d1 + (x * d0)));
                  })
```

It did become more apparent, however, that we are both reading from + writing to `result.data<T>()` in succession. This code does make it a bit more obvious that it is a safe, though, since we read from and then write to the same value.

_*note*_: no guarantee this code will compile as-is.

Reply:
Where do you read from and write to the same data?

Reply:
Oh, in the permutation iterators. Hmm, not obvious to me that this is safe...

Reply:
The "result" input permutates (maps) to the data in each column, zips, gets passed to the callback, and the return value is written back to the "result" as output. Essentially, `result[x] = f(result[x], ...)`, so it's a safe operation, albeit not obvious.